### ISSUE TITLE
bug:.net7.0 nativeaot cannot load weight.

### ISSUE THREAD

**Issue by williamlzw** [dotnet/TorchSharp#934]
vs2022, .net7.0, torchshapr0.99.3
Nativeaot can work normally, but cannot load weights
![image](https://user-images.githubusercontent.com/11348670/220804129-a44f3815-822a-49c4-a831-652921eaa095.png)

================================
begin buildmodel
begin loadmodel:D:/microocr.net/data/MicroBound_GEN9_nh64_depth2_epoch20.pth
Unhandled Exception: System.ArgumentException: Mismatched state_dict sizes: expected 0, but found 92 entries.
   at TorchSharp.torch.nn.Module.load(BinaryReader, Boolean, IList`1) + 0x170
   at TorchSharp.torch.nn.Module.load(String, Boolean, IList`1) + 0xbb
   at MicroBound.Program.test() + 0xf5
   at MicroBound!<BaseAddress>+0x3f11d3

**Comment by williamlzw** [comment]
![image](https://user-images.githubusercontent.com/11348670/220804210-6601e143-7ee7-4d6c-9f79-0a0ac992cf74.png)
![image](https://user-images.githubusercontent.com/11348670/220804248-a0834e3f-75c7-4777-b819-316fc0c88137.png)

**Comment by williamlzw** [comment]
![image](https://user-images.githubusercontent.com/11348670/220804607-7bad05ab-a668-48ab-b622-39bb3d48c2e9.png)

**Comment by NiklasGustafsson** [comment]
@williamlzw -- you'll have to show more of the model code in order for us to help you.

**Comment by williamlzw** [comment]
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static TorchSharp.torch.nn;
using static TorchSharp.torch;
using TorchSharp.Modules;
using TorchSharp;

namespace MicroBound
{
    public class ConvBNACT : Sequential
    {
        public ConvBNACT(long inChannels, long outChannels, long kernelSize, long stride = 1, long padding = 0, long groups = 1) : base(
            Conv2d(inChannels, outChannels, kernelSize, stride, padding, dilation: 1, groups: groups, bias: false),
            BatchNorm2d(outChannels),
            GELU())
        {
        }
    }

    public class MicroBlock : Module<Tensor, Tensor>
    {
        private readonly ConvBNACT conv1;
        private readonly ConvBNACT conv2;

        public MicroBlock(long nh) : base("MicroBlock")
        {
            conv1 = new ConvBNACT(nh, nh, 1);
            conv2 = new ConvBNACT(nh, nh, 3, 1, 1, nh);
            RegisterComponents();
        }

        public override Tensor forward(Tensor input)
        {
            var x = conv1.forward(input);
            x = x + conv2.forward(x);
            return x;
        }
    }

    public class MicroStage : Sequential
    {
        public MicroStage(int depth, long nh) : base(Enumerable.Range(0, depth).Select(_ => new MicroBlock(nh)).ToArray())
        {
        }
    }

    public class MLP : Sequential
    {
        public MLP(long inputDim, long hiddenDim) : base(
            Linear(inputDim, hiddenDim),
            GELU(),
            Linear(hiddenDim, inputDim)
            )
        {
        }
    }

    public class MLPBlock : Module<Tensor, Tensor>
    {
        private readonly LayerNorm layer_norm1;
        private readonly LayerNorm layer_norm2;
        private readonly LayerNorm layer_norm3;
        private readonly MLP mlp1;
        private readonly MLP mlp2;
        private readonly Conv2d conv;

        public MLPBlock(long inputDim, long hiddenDim) : base("MLPBlock")
        {
            layer_norm1 = LayerNorm(inputDim);
            mlp1 = new MLP(inputDim, hiddenDim);
            layer_norm2 = LayerNorm(inputDim);
            conv = Conv2d(inputDim, inputDim, 3, stride: 1, padding: 1, groups: inputDim, bias: false);
            layer_norm3 = LayerNorm(inputDim);
            mlp2 = new MLP(inputDim, hiddenDim);
            RegisterComponents();
        }

        public override Tensor forward(Tensor input)
        {
            var x = input.permute(0, 3, 2, 1);
            x = layer_norm1.forward(x);
            x = x + mlp1.forward(x);
            x = layer_norm2.forward(x);
            x = x.permute(0, 3, 2, 1);
            x = conv.forward(x);
            x = x.permute(0, 3, 2, 1);
            x = layer_norm3.forward(x);
            x = x + mlp2.forward(x);
            x = x.permute(0, 3, 2, 1);
            return x;
        }
    }

    public class MLPStage : Sequential
    {
        public MLPStage(int depth, long inputDim, long hiddenDim) : base(Enumerable.Range(0, depth).Select(_ => new MLPBlock(inputDim, hiddenDim)).ToArray())
        {
        }
    }

    public class Tokenizer : Sequential
    {
        public Tokenizer(long inputChannels, long hiddenDim, long outDim) : base(
            new ConvBNACT(inputChannels, hiddenDim / 2, 3, 2, 1),
            new ConvBNACT(hiddenDim / 2, hiddenDim / 2, 3, 1, 1),
            new ConvBNACT(hiddenDim / 2, outDim, 3, 1, 1),
            MaxPool2d(3, 2, 1))
        {
        }
    }

    public class Decoder : Module<Tensor, Tensor>
    {
        private readonly Sequential deconv_layers;
        private long _inChannels;
        private Sequential MakeDeconvLayer(long numLayers, List<long> numFilters, List<long> numKernels)
        {
            ModuleList<Module<Tensor, Tensor>> layer = new ModuleList<Module<Tensor, Tensor>>();
            for (int i = 0; i < numLayers; i++)
            {
                var kernel = numKernels[i];
                var planes = numFilters[i];
                layer.append(ConvTranspose2d(_inChannels, planes, kernel, 1, 1, 0, bias: false));
                layer.append(BatchNorm2d(planes));
                layer.append(GELU());
                _inChannels = planes;
            }
            return Sequential(layer);
        }
        public Decoder(long inChannels = 3) : base("Decoder")
        {
            _inChannels = inChannels;
            deconv_layers = MakeDeconvLayer(3, new List<long>(new long[] { 64, 64, 64 }), new List<long>(new long[] { 3, 3, 3 }));
            RegisterComponents();
        }

        public override Tensor forward(Tensor input)
        {
            return deconv_layers.forward(input);
        }
    }

    public class MicroBoundNet : Module<Tensor, Tensor>
    {
        private readonly Tokenizer embed;
        private readonly MicroStage micro_stages;
        private readonly MLPStage conv_stages;
        private readonly Flatten flatten;
        private readonly Decoder upsample;
        private readonly Linear fc;
        private readonly Dropout dropout;
        private readonly long _inputShape;
        public MicroBoundNet(long inputChannels = 3, long nh = 64, int depth = 2, long inputShape = 512) : base("MicroBoundNet")
        {
            embed = new Tokenizer(inputChannels, nh, nh);
            micro_stages = new MicroStage(depth, nh);
            conv_stages = new MLPStage(depth, nh, nh);
            flatten = Flatten(1, -1);
            upsample = new Decoder(nh);
            dropout = Dropout(0.1);
            fc = Linear(64, 8);
            _inputShape = inputShape;
            RegisterComponents();
        }

        public override Tensor forward(Tensor input)
        {
            using var _ = NewDisposeScope();
            var x = embed.forward(input);
            x = micro_stages.forward(x);
            x = conv_stages.forward(x);
            x = upsample.forward(x);
            x = nn.functional.adaptive_avg_pool2d(x, 1);
            x = flatten.forward(x);
            x = dropout.forward(x);
            x = fc.forward(x);
            x = x * _inputShape;
            return x.MoveToOuterDisposeScope();
        }
    }
}

**Comment by NiklasGustafsson** [comment]
And which module class is it that you are are loading?

**Comment by williamlzw** [comment]
var model = BuildModel(cfg.InChannels, cfg.Nh, cfg.Depth, cfg.InputShape);
            if (cfg.ModelPath != null)
            {
                LoadModel(cfg.ModelPath, model);
            }


  public static void LoadModel(string modelPath, Module<Tensor, Tensor> model)
        {
            model.load(modelPath);
        }

        public static Module<Tensor, Tensor> BuildModel(int inChannels, int nh, int depth, long inputShape)
        {
            return new MicroBoundNet(inChannels, nh, depth, inputShape);
        }

**Comment by williamlzw** [comment]
var model = BuildModel(cfg.InChannels, cfg.Nh, cfg.Depth, cfg.InputShape);
            if (cfg.ModelPath != null)
            {
                LoadModel(cfg.ModelPath, model);
            }


  public static void LoadModel(string modelPath, Module<Tensor, Tensor> model)
        {
            model.load(modelPath);
        }

        public static Module<Tensor, Tensor> BuildModel(int inChannels, int nh, int depth, long inputShape)
        {
            return new MicroBoundNet(inChannels, nh, depth, inputShape);
        }

**Comment by williamlzw** [comment]
Other models cannot be loaded in nativeaot mode.

**Comment by williamlzw** [comment]
Is the reading path text string of nativeaot compiled ansi characters? Cannot be loaded by torchsharp.

**Comment by NiklasGustafsson** [comment]
What is the result of this:

```C#
var model = BuildModel(cfg.InChannels, cfg.Nh, cfg.Depth, cfg.InputShape);

var size = model.state_dict().Count;
```

**Comment by williamlzw** [comment]
begin buildmodel
model size:0 /// 0
begin loadmodel:D:/microocr.net/data/MicroBound_GEN9_nh64_depth2_epoch20.pth
Unhandled Exception: System.ArgumentException: Mismatched state_dict sizes: expected 0, but found 92 entries.
   at TorchSharp.torch.nn.Module.load(BinaryReader, Boolean, IList`1) + 0x170
   at TorchSharp.torch.nn.Module.load(String, Boolean, IList`1) + 0xbb
   at MicroBound.Program.test() + 0x1bc
   at MicroBound!<BaseAddress>+0x3fc2f3

**Comment by williamlzw** [comment]
The model size is 92 and runs using. net. Use nativeaot to run size 0
===========
begin buildmodel
model size:92
begin loadmodel:D:/microocr.net/data/MicroBound_GEN9_nh64_depth2_epoch20.pth
begin readimg:d:\dataset\gen9\test\00000.jpg
begin forward
end forward

**Comment by NiklasGustafsson** [comment]
Okay, that's the problem -- it's not the load, it's the initialization of the model that's going wrong for some reason. 

TorchSharp has not been tested with AOT.

**Comment by williamlzw** [comment]
Can you fix the problem of aot compilation, saving and loading weights? Now the model is trained and reasoned normally, but the weight cannot be loaded and saved. Thank you.

**Comment by NiklasGustafsson** [comment]
@tarekgh -- any thoughts and/or insights?

**Comment by tarekgh** [comment]
@NiklasGustafsson AOT is not friendly with the reflection. If someone can look at the failing code path and figure out the area that the reflection is used there, that can help try to find what is the best way to mitigate that for AOT.

**Comment by williamlzw** [comment]
https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/reflection-free-mode.md
https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/reflection-in-aot-mode.md

**Comment by williamlzw** [comment]
![image](https://user-images.githubusercontent.com/11348670/223010491-e7dd6945-c0ef-46e8-9e78-2b6058fff9a9.png)

**Comment by williamlzw** [comment]
![image](https://user-images.githubusercontent.com/11348670/223010639-b3651dfa-bb22-409c-93ff-9b4fea9a917d.png)

**Comment by NiklasGustafsson** [comment]
@williamlzw -- thanks for identifying the spot where reflection is used. This is necessary in order to make the definition of custom modules straightforward, especially the loading and storing of state dictionaries, but also the registration of components when a module is created, and the conversion of modules to a different element type or moving modules to another processor. 

If AOT is incompatible with the use of FieldInfo, then that's very unfortunate. We'll have to add some visible comment to the readme and other documentation.

**Comment by tarekgh** [comment]
Is it possible you can try providing the [rd.xml](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/reflection-in-aot-mode.md#providing-hints-to-the-compiler-externally) for the Module class and find out if this can help?

**Comment by williamlzw** [comment]
Net8.0 is expected to support method reflection in the near future.
